Core/Src/main.c

```diff
@@ -134,16 +142,45 @@ void SystemClock_Config(void)
   {
     Error_Handler();
   }
-  PeriphClkInit.PeriphClockSelection = RCC_PERIPHCLK_USART2;
-  PeriphClkInit.Usart2ClockSelection = RCC_USART2CLKSOURCE_PCLK1;
-  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInit) != HAL_OK)
-  {
-    Error_Handler();
-  }
 }
 
 /* USER CODE BEGIN 4 */
+uint16_t readADC (uint8_t channel){
+
+
+	uint8_t receivedData[2] = {0};
+
+    uint8_t start_bit = 0x01;
+    uint8_t sglDiff = 1;      // singleended
+    uint8_t D2 = (channel >> 2) & 0x01;  // extract bit 2 of channel
+    uint8_t D1 = (channel >> 1) & 0x01;  // bit 1
+    uint8_t D0 = channel & 0x01;
+
+    uint8_t transmitData[3];
+    transmitData[0] = start_bit; //start
+    transmitData[1] = (sglDiff << 7) | (D2 << 6) | (D1 << 5) | (D0 << 4); //combine
+    transmitData[2] = 0x00;      //dummy byte
 
+	HAL_GPIO_WritePin(CS_PORT, CS_PIN, GPIO_PIN_RESET); //start transmission going low
+	HAL_SPI_TransmitReceive(&hspi1, transmitData, receivedData, 3, HAL_MAX_DELAY);
+	HAL_GPIO_WritePin(CS_PORT, CS_PIN, GPIO_PIN_SET);
+
+	uint16_t result = ((receivedData[1] & 0x0F) << 8) | receivedData[2];
```

Comment:
I think the mask that is compared with receivedData[1] is better to be 0x03, as only the last two bits in that data byte are valid data. The first few bits can be random and affect the correct result.